### ISSUE TITLE
Feature request: virtual oneof property

### ISSUE THREAD

**Issue by paskozdilar** [stephenh/ts-proto#1099]
The `protobuf.js` library provides [an option](https://github.com/protobufjs/protobuf.js/blob/master/index.d.ts#L1740) `oneofs` for setting virtual oneof fields that contain name of the set property:

```typescript
    /** Includes virtual oneof properties set to the present field's name, if any */
    oneofs?: boolean;
```

Similar option already exists in the generated pbjs.d.ts files in integration tests, e.g. in `oneof-properties`:

```typescript
export namespace oneof {
    // ...
    class PleaseChoose implements IPleaseChoose {
        // ...
        public choice?: ("aNumber"|"aString"|"aMessage"|"aBool"|"bunchaBytes"|"anEnum");
        public eitherOr?: ("either"|"or"|"thirdOption");
        // ...
    }
    // ...
}
```

It would be ideal to have generated interface actually be a flat union type with a virtual parameter, so TypeScript can check for field existence based on virtual field value, but my [naive attempt](https://github.com/paskozdilar/ts-proto/blob/feature/oneof-unions-virtual/integration/oneof-unions-virtual/oneof.ts) has caused combinatorial explosion of union types related to `Exact` and `DeepPartial` which are too complicated for me to tackle at the moment.

Having just a virtual parameter is a weaker - but still useful - feature.

I am planning to do the work once I get more comfortable with the codebase. Any pointers would be welcome :)

**Comment by stephenh** [comment]
Hi @paskozdilar ! Just sanity checking, but have you seen our existing oneof handling, i.e.:

https://github.com/stephenh/ts-proto?tab=readme-ov-file#oneof-handling

And specifically the `oneof=unions-value` option?

It outputs code like:

```
interface YourMessage {
  eitherField?: { $case: "field_a"; value: string } | { $case: "field_b"; value: string };
}
```

Which seems very similar to this request, but I can't 100% tell ... is there something about the `unions-value` output (or older `unions` output) that is different from your proposal/what you'd like to see?

**Comment by paskozdilar** [comment]
Yes, it's very similar, but unfortunately the `oneof=unions-value` is not compatible with default [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway) setup, which uses flat representation like `ts-proto` default.

However, the problem is twofold:

1. It's impossible to know which fields belong to which `oneof` from `ts-proto` generated types alone
2. It's impossible to assert that we've checked for all types available in a OneOf.

---

The first problem could be solved by the virtual property field that would contain the name of the set oneof property, e.g.:

```protobuf
message Foo {
  oneof either_field { string field_a = 1; string field_b = 2; }
}
```

would generate the following:

```typescript
export interface Foo {
  either_field: 'field_a' | 'field_b' | undefined;
  field_a: string | undefined;
  field_b: string | undefined;
}

// we can now do this:
function fn(arg: Foo) {
    switch (arg.either_field) {
        case 'field_a':
            // ... handle field a
        case 'field_b':
            // ... handle field b
        default:
            // compiler will fail here if we haven't exhausted all non-undefined cases
            let _: undefined;
            _ = args.either_field;
    }
}
```

That wouldn't give complete type-safety, but would still increase convenience.

In the current implementation of flat-fields, all oneofs have to be checked manually, and one has to look at the proto file itself to know which field belongs to which oneof.

**Comment by andrei-kvl** [comment]
@stephenh 
Hi!
Are there any updates?

**Comment by andrei-kvl** [comment]
This behavior corresponds to the **oneofs** option in **proto-loader**. It would be convenient to have such a field for consistency with the **oneofs** option.

https://github.com/grpc/grpc-node/tree/master/packages/proto-loader

**Comment by paskozdilar** [comment]
Closing this for now as I haven't found the time to work on it, and probably won't in the near future.
